{% extends "defindex.html" %}

{% block tables %} 

<div style='padding-right:30px;
            padding-left:30px;
            font-size:100%;
            max-width:800px;'>

<div >
<p style='text-align:left;
            line-height:160%;
            color:#555;'>
ETE is a Python programming toolkit that assists in the <i>automated
manipulation, analysis and visualization of phylogenetic trees</i>.

It provides a wide range of tree handling options, node annotation features and
specialized features to deal with <i>phylogenetic trees</i> (i.e automatic
orthology and paralogy detection, phylostratigraphy, tree reconciliation,
etc). ETE implements also an interactive <i>tree visualization</i> system based on a a
highly customizable tree drawing engine (PDF and SVG tree images).

Although ETE is developed as a tool for phylogenetic analysis,
it is also used to handle other types of hierarchical trees (i.e. clustering results). 
</p>
</div>

<table class="contentstable" align='center'><tr>
    
    <td width="50%">
        <p class="biglink"><a class="biglink" href="{{ pathto("tutorial/index") }}">Python API Tutorial</a><br/><br/>
            <!-- <span class="linkdescr">How to use the different modules of the library</span> -->
        </p>

        <p class="biglink"><a class="biglink" href="{{ pathto("reference/index") }}">Python API Reference</a><br/><br/>
            <!-- <span class="linkdescr">An index of all ETE functions</span> -->
        </p>

        
        <p class="biglink"><a class="biglink" href="http://etetoolkit.org/documentation/tools/">Phylogenomic tools</a><br/><br/>
            <!-- <span class="linkdescr">Tools to visualize, compare, and handle trees</span> -->
        </p>        
        
    </td><td width="50%">
        <!-- <p class="biglink"><a class="biglink" href="{{ pathto("changelog/changelog2.3") }}">What's new in ETE {{ version }}?</a><br/>
        <a href="{{ pathto("changelog/index") }}">Previous versions</a>
        </p> -->
        
        <p class="biglink"><a class="biglink" href="http://etetoolkit.org/download/">Setup and Install</a><br/><br/>
            <!-- <span class="linkdescr">ETE install and requirements</span> -->
        </p>
        
        <p class="biglink"><a class="biglink" href="http://github.org/jhcepas/ete/issues/">Contributing</a><br/><br/>
            <!-- <span class="linkdescr">Feature requests, bug reports, etc.</span> -->
        </p>
        
    </td></tr>
</table>

<div style='font-size:100%;font-family:"Helvetica Neue",Helvetica,Arial,sans-serif;line-height:110%;color:#666;'>
    <div style='margin:auto;border-top:1px solid blue; border-bottom:1px solid blue; padding-top:10px; padding-bottom:10px; width:85%;font-size:88%;'>
        If you use ETE to analyze, process or visuallize results in a published
        work, please do not forget to support the project by citing:
        <br><br>
        <span style="font-size:80%;">
            <i>Jaime Huerta-Cepas, Joaquín Dopazo and Toni Gabaldón.</i> ETE: a python Environment for Tree Exploration. BMC Bioinformatics 2010, 11:24. 
                <a alt="link to citation reference" title='link to citation reference' target='_black' href='http://www.biomedcentral.com/1471-2105/11/24'>doi:10.1186/1471-2105-11-24</a>
            
        </span>
    </div>
</div>

  
  <div class="note" style="border:0px;text-align:center;background-color:#ffffff">
      <!-- <a href="https://github.com/jhcepas/ete/tree/2.3/examples"> Check more examples </a>
      <br> -->
    <a href="_downloads/ETE.pdf"> Download this documentation as PDF. </a>

  </div>

  <div class="note" style="text-align:center;background-color:#D0ECa2">
    More info and examples are available at <a href="http://etetoolkit.org" target="_blank">http://etetoolkit.org </a>
  </div>
</div>

{% endblock %}
